### PR TITLE
feat: add multivariate anomaly detection

### PR DESCRIPTION
**Pull request by serena-ruan**

### CODE REVIEW

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/FormOntologyLearner.scala [hunk 1]
```diff
@@ -53,21 +53,23 @@ class FormOntologyLearner(override val uid: String) extends Estimator[FormOntolo
...
-    val mergedSchema = dataset.toDF()
-      .select(col(getInputCol))
-      .map(extractOntology(fromRow))(Encoders.kryo[StructType])
-      .reduce(combine _)
-
-    new FormOntologyTransformer()
-      .setInputCol(getInputCol)
-      .setOutputCol(getOutputCol)
-      .setOntology(mergedSchema)
+    logFit({
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/FormOntologyLearner.scala:56]: Thanks :)

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+                               location: URI): Option[HTTPResponseData] = {
+    val get = new HttpGet()
+    get.setURI(location)
+    key.foreach(get.setHeader("Ocp-Apim-Subscription-Key", _))
+    get.setHeader("User-Agent", s"synapseml/${BuildInfo.version}${HeaderValues.PlatformInfo}")
+    val resp = convertAndClose(sendWithRetries(client, get, getBackoffs))
+    get.releaseConnection()
+    val fields = IOUtils.toString(resp.entity.get.content, "UTF-8").parseJson.asJsObject.fields
+    var status = "none"
+    if (fields.keySet.contains("modelInfo")) {
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala:99]: var can be replaced with conditional equality. Ex:  val foo = if(..){1}else{2}

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+  def setSlidingWindowCol(v: String): this.type = setVectorParam(slidingWindow, v)
+
+  def getSlidingWindow: Int = getScalarParam(slidingWindow)
+
+  def getSlidingWindowCol: String = getVectorParam(slidingWindow)
+
+  val alignMode = new ServiceParam[String](this, "alignMode", "An optional field, indicates how " +
+    "we align different variables into the same time-range which is required by the model.{Inner, Outer}", {
+    case Left(s) => Set("Inner", "Outer")(s)
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala:166]: nit: Might want to make this case insensitive

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+  def setAlignModeCol(v: String): this.type = setVectorParam(alignMode, v)
+
+  def getAlignMode: String = getScalarParam(alignMode)
+
+  def getAlignModeCol: String = getVectorParam(alignMode)
+
+  val fillNAMethod = new ServiceParam[String](this, "fillNAMethod", "An optional field, indicates how missed " +
+    "values will be filled with. Can not be set to NotFill, when alignMode is Outer.{Previous, Subsequent," +
+    " Linear, Zero, Fixed}", {
+    case Left(s) => Set("Previous", "Subsequent", "Linear", "Zero", "Fixed")(s)
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala:181]: likewise here

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+
+  def setFillNAMethod(v: String): this.type = setScalarParam(fillNAMethod, v)
+
+  def setFillNAMethodCol(v: String): this.type = setVectorParam(fillNAMethod, v)
+
+  def getFillNAMethod: String = getScalarParam(fillNAMethod)
+
+  def getFillNAMethodCol: String = getVectorParam(fillNAMethod)
+
+  val paddingValue = new ServiceParam[Int](this, "paddingValue", "optional field, only be useful" +
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala:193]:
```suggestion
  val paddingValue = new ServiceParam[Int](this, "paddingValue", "optional field,  is only useful" +
```

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+      }
+    } else {
+      response
+    }
+  }
+}
+
+object MultivariateAnomalyModel extends ComplexParamsReadable[MultivariateAnomalyModel] with Serializable
+
+class MultivariateAnomalyModel(override val uid: String) extends Estimator[DetectMultivariateAnomaly]
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala:142]: Maybe we can rename this to MultivariateAnomalyEstimator so as not to confuse with the model that it spits out

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetectorSchemas.scala [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                      displayName: Option[String])
+
+object MAMResponse extends SparkBindings[MAMResponse]
+
+case class MAMResponse(modelId: String,
+                       createdTime: String,
+                       lastUpdatedTime: String,
+                       modelInfo: MAMModelInfo)
+
+case class MAMModelInfo(slidingWindow: Option[Int],
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetectorSchemas.scala:64]: nit: Could you put a comment in this file that describes what the MAM and DMA acronyms stand for?

File: cognitive/src/test/scala/com/microsoft/azure/synapse/ml/cognitive/split1/MultivariateAnamolyDetectionSuite.scala [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    def prep(df: DataFrame) = {
+      df.select("source")
+    }
+
+    super.assertDFEq(prep(df1), prep(df2))(eq)
+  }
+
+  def mad: MultivariateAnomalyModel = new MultivariateAnomalyModel()
+    .setSubscriptionKey(anomalyKey)
+    .setLocation("westus2")
```
Comment (mhamilton723) [inline cognitive/src/test/scala/com/microsoft/azure/synapse/ml/cognitive/split1/MultivariateAnamolyDetectionSuite.scala:120]: Do they have this service in eastus?

**Review by mhamilton723 — CHANGES_REQUESTED** [review]
Looks awesome! Only a few minor things

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala (hunk 2 above)
Comment (serena-ruan) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala]: Done! Good to know this!

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala (hunk 3 above)
Comment (serena-ruan) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala]: Done.

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala (hunk 4 above)
Comment (serena-ruan) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala]: Done.

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala (hunk 5 above)
Comment (serena-ruan) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala]: Done.

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala (hunk 6 above)
Comment (serena-ruan) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetection.scala]: Sure. Done.

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetectorSchemas.scala (hunk 7 above)
Comment (serena-ruan) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/MultivariateAnomalyDetectorSchemas.scala]: Yes just added.

File: cognitive/src/test/scala/com/microsoft/azure/synapse/ml/cognitive/split1/MultivariateAnamolyDetectionSuite.scala (hunk 8 above)
Comment (serena-ruan) [reply cognitive/src/test/scala/com/microsoft/azure/synapse/ml/cognitive/split1/MultivariateAnamolyDetectionSuite.scala:120]: Yes they do, but our existing one is in westus2 so I just use it.

**Review by mhamilton723 — APPROVED** [review]

**Pull request merged**